File: prototype/internal/blocks/blocks.go
// Package blocks implements the in-memory store and DTOs for notebook blocks.
//
// Mirrors the Rust `src/blocks.rs` and Python `notebook_engine.blocks` contracts
// so that all three cascade implementations expose identical JSON shapes.
package blocks

import (
	"errors"
	"sync"

	"github.com/google/uuid"
)

// Block is the persisted shape returned by all block endpoints.
type Block struct {
	ID     string `json:"id"`
	MinCol uint32 `json:"min_col"`
	MaxCol uint32 `json:"max_col"`
	MinRow uint32 `json:"min_row"`
	MaxRow uint32 `json:"max_row"`
}

// BlockCreate is the POST /api/blocks request body.
//
// The payload accepts EITHER fully-specified bounds (min_col/max_col/min_row/max_row)
// OR start/end point pairs (start_col/start_row/end_col/end_row). Mixing shapes
// or providing an incomplete shape is rejected.
type BlockCreate struct {
	MinCol *uint32 `json:"min_col,omitempty"`
	MaxCol *uint32 `json:"max_col,omitempty"`
	MinRow *uint32 `json:"min_row,omitempty"`
	MaxRow *uint32 `json:"max_row,omitempty"`

	StartCol *uint32 `json:"start_col,omitempty"`
	StartRow *uint32 `json:"start_row,omitempty"`
	EndCol   *uint32 `json:"end_col,omitempty"`
	EndRow   *uint32 `json:"end_row,omitempty"`
}

// Bounds holds validated (min_col, max_col, min_row, max_row).
type Bounds struct {
	MinCol, MaxCol, MinRow, MaxRow uint32
}

// ToBounds validates the request shape and normalizes start/end into bounds.
func (b BlockCreate) ToBounds() (Bounds, error) {
	hasBounds := b.MinCol != nil || b.MaxCol != nil || b.MinRow != nil || b.MaxRow != nil
	hasPoints := b.StartCol != nil || b.StartRow != nil || b.EndCol != nil || b.EndRow != nil

	if hasBounds && hasPoints {
		return Bounds{}, errors.New("provide either bounds fields or start/end fields, not both")
	}
	if !hasBounds && !hasPoints {
		return Bounds{}, errors.New("provide one complete block shape: bounds or start/end")
	}

	if hasBounds {
		if b.MinCol == nil || b.MaxCol == nil || b.MinRow == nil || b.MaxRow == nil {
			return Bounds{}, errors.New("bounds shape requires min_col, max_col, min_row, max_row")
		}
		if *b.MinCol > *b.MaxCol {
			return Bounds{}, errors.New("min_col must be <= max_col")
		}
		if *b.MinRow > *b.MaxRow {
			return Bounds{}, errors.New("min_row must be <= max_row")
		}
		return Bounds{*b.MinCol, *b.MaxCol, *b.MinRow, *b.MaxRow}, nil
	}

	if b.StartCol == nil || b.StartRow == nil || b.EndCol == nil || b.EndRow == nil {
		return Bounds{}, errors.New("start/end shape requires start_col, start_row, end_col, end_row")
	}
	return Bounds{
		MinCol: min(*b.StartCol, *b.EndCol),
		MaxCol: max(*b.StartCol, *b.EndCol),
		MinRow: min(*b.StartRow, *b.EndRow),
		MaxRow: max(*b.StartRow, *b.EndRow),
	}, nil
}

// Store is the interface implemented by InMemoryStore.
type Store interface {
	List() []Block
	Create(b Bounds) Block
	Delete(id string) bool
	Clear()
}

// InMemoryStore is a goroutine-safe Store backed by a map.
type InMemoryStore struct {
	mu     sync.RWMutex
	blocks map[string]Block
}

// NewInMemoryStore constructs an empty store.
func NewInMemoryStore() *InMemoryStore {
	return &InMemoryStore{blocks: make(map[string]Block)}
}

// List returns all blocks in unspecified order.
func (s *InMemoryStore) List() []Block {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]Block, 0, len(s.blocks))
	for _, b := range s.blocks {
		out = append(out, b)
	}
	return out
}

// Create inserts a new block with a generated UUIDv4 id.
func (s *InMemoryStore) Create(b Bounds) Block {
	block := Block{
		ID:     uuid.NewString(),
		MinCol: b.MinCol,
		MaxCol: b.MaxCol,
		MinRow: b.MinRow,
		MaxRow: b.MaxRow,
	}
	s.mu.Lock()
	s.blocks[block.ID] = block
	s.mu.Unlock()
	return block
}

// Delete removes a block by id; returns false if not found.
func (s *InMemoryStore) Delete(id string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.blocks[id]; !ok {
		return false
	}
	delete(s.blocks, id)
	return true
}

// Clear drops all blocks.
func (s *InMemoryStore) Clear() {
	s.mu.Lock()
	s.blocks = make(map[string]Block)
	s.mu.Unlock()
}

// CreateForGrid enforces the margin invariant (min_col >= marginCols).
func CreateForGrid(store Store, payload BlockCreate, marginCols uint32) (Block, error) {
	bounds, err := payload.ToBounds()
	if err != nil {
		return Block{}, err
	}
	if bounds.MinCol < marginCols {
		return Block{}, errors.New("min_col must be >= margin_cols")
	}
	return store.Create(bounds), nil
}

File: rust-prototype/src/blocks.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::RwLock;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Block {
    pub id: String,
    pub min_col: u32,
    pub max_col: u32,
    pub min_row: u32,
    pub max_row: u32,
}

#[derive(Debug, Deserialize)]
pub struct BlockCreate {
    pub min_col: Option<u32>,
    pub max_col: Option<u32>,
    pub min_row: Option<u32>,
    pub max_row: Option<u32>,
    pub start_col: Option<u32>,
    pub start_row: Option<u32>,
    pub end_col: Option<u32>,
    pub end_row: Option<u32>,
}

impl BlockCreate {
    pub fn to_bounds(&self) -> Result<(u32, u32, u32, u32), &'static str> {
        let has_bounds = self.min_col.is_some() || self.max_col.is_some() || self.min_row.is_some() || self.max_row.is_some();
        let has_points = self.start_col.is_some() || self.start_row.is_some() || self.end_col.is_some() || self.end_row.is_some();

        if has_bounds && has_points {
            return Err("Provide either bounds fields or start/end fields, not both");
        }
        if !has_bounds && !has_points {
            return Err("Provide one complete block shape: bounds or start/end");
        }

        if has_bounds {
            let min_col = self.min_col.ok_or("Bounds shape requires min_col, max_col, min_row, max_row")?;
            let max_col = self.max_col.ok_or("Bounds shape requires min_col, max_col, min_row, max_row")?;
            let min_row = self.min_row.ok_or("Bounds shape requires min_col, max_col, min_row, max_row")?;
            let max_row = self.max_row.ok_or("Bounds shape requires min_col, max_col, min_row, max_row")?;

            if min_col > max_col || min_row > max_row {
                return Err("min_col must be <= max_col and min_row must be <= max_row");
            }
            Ok((min_col, max_col, min_row, max_row))
        } else {
            let start_col = self.start_col.ok_or("Start/end shape requires start_col, start_row, end_col, end_row")?;
            let start_row = self.start_row.ok_or("Start/end shape requires start_col, start_row, end_col, end_row")?;
            let end_col = self.end_col.ok_or("Start/end shape requires start_col, start_row, end_col, end_row")?;
            let end_row = self.end_row.ok_or("Start/end shape requires start_col, start_row, end_col, end_row")?;

            Ok((
                start_col.min(end_col),
                start_col.max(end_col),
                start_row.min(end_row),
                start_row.max(end_row),
            ))
        }
    }
}

pub struct BlockStore {
    blocks: RwLock<HashMap<String, Block>>,
}

impl BlockStore {
    pub fn new() -> Self {
        Self {
            blocks: RwLock::new(HashMap::new()),
        }
    }

    pub fn list(&self) -> Vec<Block> {
        let map = self.blocks.read().unwrap();
        map.values().cloned().collect()
    }

    pub fn create(&self, min_col: u32, max_col: u32, min_row: u32, max_row: u32) -> Block {
        let id = Uuid::new_v4().to_string();
        let block = Block {
            id: id.clone(),
            min_col,
            max_col,
            min_row,
            max_row,
        };
        self.blocks.write().unwrap().insert(id, block.clone());
        block
    }

    pub fn delete(&self, id: &str) -> bool {
        self.blocks.write().unwrap().remove(id).is_some()
    }

    pub fn clear(&self) {
        self.blocks.write().unwrap().clear();
    }
}
